Cxc 178 undo redo actions

hannahwagner1402 opened the pull request:
Undo/Redo functionality has been added 

Transforming, deleting, and adding the assets and text should be able to be undone/ redone
The undo/ redo buttons are disabled until there is something to undo/ redo
You have a history/ redo array of 10 and as soon as you change something you can no longer redo 

Big bugs i encountered: 
- when deleting or clearing an asset I couldnt undo deleting it
- Being able to undo changes in the text was buggy
- clicking between undo and redo multiple times was very buggy for a while

Points that you are not sure you did as you should:
-The code definitely needs to be refactored a bit, but it works 
- Sometimes when undoing the assets are active and it looks a bit weird, but has no affect on the functionality.

Points that you want the reviewer to pay special attention on:
-testing all transforms and ensuring that everything works especially with deleting assets

esitsolf commented on utils/Classes/Panel.js:
```diff
@@ -22,136 +15,194 @@ export default class Panel {
+    }
+
+    redoAction() {
+        if (this._redo.length > 0) {
+            let nextState = this._redo.pop();
+            this.applyState(nextState);
+            this.history.push(nextState);
+        }
```
I don't think you need to control the redo.length here since you can only click if the button is enabled (which can only occure if the redo as something in it)

esitsolf commented on utils/Classes/Panel.js:
```diff
@@ -22,136 +15,194 @@ export default class Panel {
+    }
+
+    undo() {
+        if (this.history.length <= 1) return;
```
I don't think you need to control the history length here. This function can only be called when the button is enabled (which can only occur if history has something in it)

esitsolf reviewed (changes requested):
You did a great Job! I particully enjoy, that each panel has its own undo redo possibility.

The comments I let on the code are up to you since you know your code better than I do.

There are 2 bugs that you should try to solve before merging (as talked on discord): 
1. The text undo redo is buggy (text size not working on firefox, 2 clicks needed on chrome, text content not always updated)
2. The edition menu disappears after undoing or redoing an action, but the element is active.
=> I think this bug comes from the elementActive property of dragResizeRotate, which trigger the v-if of the editionMenu (as talked in discord)
=> an other option could be that no element is selected after redoing or undoing.

Vanessi-K commented on components/ComicPanels.vue:
```diff
@@ -39,6 +40,7 @@
         <swiper-slide v-for="(panel, index) in comic.getPage(0).getStrip(0).panels" :key="index">
             <WrapperCanvas
```
I think you should be good if you just put the key already here instead of further passing the refresh-count. I don't think that it makes sense, that the `WrapperCanvas` would handle it's refresh internally.

Vanessi-K reviewed (changes requested):
1. When I double click on a text to edit it I think each click seperately is added to the sack of undo/redo, although in this case only the real action of changing than stuff (text, font-size) in the pop-up should be considered as something that is added to undo/redo.
2. I think only the final change of the font-size should be added to undo/redo. Or in general the state when closing the text-popup. The in between steps of the font-size change do not need to be added to undo/redo. It also is not for the typing of text.
  - I think changing font-size and text (so all changes from the TextEditor) should be one change instead of seperate, since you never see them seperate on the editor.

hannahwagner1402 commented on utils/Classes/Panel.js:
```diff
@@ -1,17 +1,10 @@
+import Text from './Text';
+
 export default class Panel {
-    /** @type {String} */
-    _border;
-    /** @type {Map<String, ElementDS>} */
-    _elements;
-    /** @type {Number} */
-    _width;
-    /** @type {Number} */
-    _height;
-
-    /**
-     * @param {Number} width
-     * @param {String} border
-     */
```
ah i think i did this on accident when i was trying to refactor some things, i will add them back to keep everthing consistent, sorry!

Vanessi-K commented on components/TextEditor.vue:
```diff
@@ -1,52 +1,58 @@
 <script setup>
+    import { ref, onMounted, watch } from 'vue';
+    import { useComicStore } from '#imports';
```
These should all be auto imported. You don't need the inputs.

Vanessi-K commented on components/DragResizeRotate.vue:
```diff
@@ -161,7 +163,6 @@
         :h="h"
         :eId="eId"
         class-name-active="element--active"
-        class-name-handle="handle-large"
```
I think this should still exist. There is properly an error in merging.

Vanessi-K commented on components/DragResizeRotate.vue:
```diff
@@ -204,7 +205,6 @@
         </div>
 
         <img
-            draggable="false"
```
I think this should still exist. There is properly an error in merging.

Vanessi-K commented on components/DragResizeRotate.vue:
```diff
@@ -229,7 +229,6 @@
         &__image {
             width: 100%;
             height: 100%;
-            user-select: none;
```
I think this should still exist. There is properly an error in merging.

Vanessi-K reviewed (changes requested):
Good job!

The functionality and docs look good now! 

The only annoying thing left would be the active-state of an element after undo/redo (which was also already mentioned by @esitsolf ).
What were your findings on that @hannahwagner1402, can it be solved?

Also please see my in-code-comments, since I assume something did go wrong in merging where new lines are now deleted.

hannahwagner1402 commented:
@Vanessi-K I fixed all the changes regarding the merge error and removed the unnecessary imports. I tried finding a solution to the issue with the elementActive property but was not able to fix the issue. Antoine and I decided it would be best to keep that as an improvement for the future as it does not interfere directly with the functionality of the undo redo.

Vanessi-K commented:
> Antoine and I decided it would be best to keep that as an improvement for the future as it does not interfere directly with the functionality of the undo redo.

I am ok with that. Please add it to our list.

Vanessi-K commented:
I have like one more request. Changes made in the layer-order over the layer-pop-up are not considered in undo/redo.
